optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,281 @@
+            relative search space are sampled by this sampler.
+
+            If :obj:`None` is specified, :class:`~optuna.samplers.RandomSampler` is used
+            as the default.
+
+            .. seealso::
+                :class:`~optuna.samplers` module provides built-in independent samplers
+                such as :class:`~optuna.samplers.RandomSampler` and
+                :class:`~optuna.samplers.TPESampler`.
+
+        warn_independent_sampling:
+            If this is :obj:`True`, a warning message is emitted when
+            the value of a parameter is sampled by using an independent sampler.
+
+            Note that the parameters of the first trial in a study are sampled via an
+            independent sampler in most cases, so no warning messages are emitted in such cases.
+
+    Raises:
+        ValueError:
+            If ``qmc_type`` is not one of 'sobol', 'halton', 'LHS' or 'OA-LHS'.
+
+    .. note::
+        Added in v2.x.0 TODO(kstoneriv3)as an experimental feature. The interface may change in
+        newer versions without prior notice.
+
+    Example:
+
+        Optimize a simple quadratic function by using :class:`~optuna.samplers.QMCSampler`.
+
+        .. testcode::
+
+            import optuna
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -1, 1)
+                y = trial.suggest_int("y", -1, 1)
+                return x ** 2 + y
+
+
+            sampler = optuna.samplers.QMCSampler()
+            study = optuna.create_study(sampler=sampler)
+            study.optimize(objective, n_trials=20)
+
+    """
+
+    def __init__(
+        self,
+        *,
+        qmc_type: str = "sobol",
+        scramble: bool = False,
+        seed: Optional[int] = None,
+        search_space: Optional[Dict[str, BaseDistribution]] = None,
+        independent_sampler: Optional[BaseSampler] = None,
+        warn_independent_sampling: bool = True,
+    ) -> None:
+        self._scramble = scramble
+        self._seed = seed
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._qmc_type = qmc_type
+        self._qmc_engine = None
+        # TODO(kstoneriv3): make sure that search_space is either None or valid search space.
+        # also make sure that it is OrderedDict
+        self._initial_search_space = search_space
+        self._warn_independent_sampling = warn_independent_sampling
+
+    def reseed_rng(self) -> None:
+
+        self._independent_sampler.reseed_rng()
+
+    def infer_relative_search_space(
+        self, study: Study, trial: FrozenTrial
+    ) -> Dict[str, BaseDistribution]:
+
+        if self._initial_search_space is not None:
+            return self._initial_search_space
+
+        past_trials = study._storage.get_all_trials(study._study_id, deepcopy=False)
+        past_trials = [t for t in past_trials if t.state in _SUGGESTED_STATES]
+        past_trials = sorted(past_trials, key=lambda t: t._trial_id)
```

Comment:
Should the key be `trial.number`?
```suggestion
        past_trials = sorted(past_trials, key=lambda t: t.number)
```